# Python制作数据分析工具

## 一 案例介绍

### 1 数据诊断的目的

- 了解特征的分布, 缺失和异常等情况
- 统计指标可直接用于数据处理

![](./picture/1-1.png)
![](./picture/1-2.png)

## 二 数据诊断整体介绍

### 2-1 依赖包

- `numpy`
- `pandas`
- `scipy`

`from __future__ import division`<br>
`import pandas as pd`<br>
`import numpy as np`<br>
`from scipy import stats`

### 2-2 数据获取与数据读取

- 数据获取: `Kaggle`
- 数据读取: `https://www.kaggle.com/c/santander-customer-satisfaction`

### 2-3 统计指标介绍

- 均值 / 中位数 / 最大值 / 最小值
- 计数类(某一个数出现的次数)
- 缺失值 / 方差等
- 分位点 / 值的频数等
 - 分位点: 数据的分布情况
 - 值的频数: 某一个数经常出现的数, 该值的比例!

## 三 数据诊断工具制作

### 3-1 描述统计-计数类

In [118]:
import pandas as pd
import numpy as np
from scipy import stats

df = pd.read_csv("./data/train.csv")
label = df["TARGET"]
df = df.drop(["ID", "TARGET"], axis=1)

#### 1 定义缺失值 `Missing Value`

In [8]:
missSet = [np.nan, 9999999999, -999999]
print(missSet)

[nan, 9999999999, -999999]


In [55]:
print(df.shape)
df.head(5)

(76020, 369)


var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0     2     23                 0.0                      0.0   
1     2     34                 0.0                      0.0   
2     2     23                 0.0                      0.0   
3     2     37                 0.0                    195.0   
4     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  imp_op_var40_ult1  \
0                      0.0                      0.0                0.0   
1                      0.0                      0.0                0.0   
2                      0.0                      0.0                0.0   
3                      0.0                      0.0                0.0   
4                      0.0                      0.0                0.0   

       ...        saldo_medio_var29_ult3  saldo_medio_var33_hace2  \
0      ...                           0.0                      0.0   
1      ...                           0.0                      0.0   
2      ...                           0.0                      0.0   
3      ...                           0.0                      0.0   
4      ...                           0.0                      0.0   

   saldo_medio_var33_hace3  saldo_medio_var33_ult1  saldo_medio_var33_ult3  \
0                      0.0                     0.0                     0.0   
1                      0.0                     0.0                     0.0   
2                      0.0                     0.0                     0.0   
3                      0.0                     0.0                     0.0   
4                      0.0                     0.0                     0.0   

   saldo_medio_var44_hace2  saldo_medio_var44_hace3  saldo_medio_var44_ult1  \
0                      0.0                      0.0                     0.0   
1                      0.0                      0.0                     0.0   
2                      0.0                      0.0                     0.0   
3                      0.0                      0.0                     0.0   
4                      0.0                      0.0                     0.0   

   saldo_medio_var44_ult3          var38  
0                     0.0   39205.170000  
1                     0.0   49278.030000  
2                     0.0   67333.770000  
3                     0.0   64007.970000  
4                     0.0  117310.979016  

[5 rows x 369 columns]

#### 2 不同值的数量 `Count distinct`

查看第一列不同值的数量

In [21]:
print("Count Distinct:", len(df.iloc[:, 0].unique()))

Count Distinct: 208


取出前三列, apply会遍历每一列并通过后面的lambda表达式统计不同值的数量

In [22]:
count_un = df.iloc[:, 0:3].apply(lambda x:len(x.unique()))
print(count_un)

var3                  208
var15                 100
imp_ent_var16_ult1    596
dtype: int64


#### 3 `Zero Values`

判断第一列有多少个0

In [23]:
print("Zero Values Sum:", np.sum(df.iloc[:, 0] == 0))

Zero Values Sum: 75


查看前三列有多少个0

In [24]:
count_zero = df.iloc[:, 0:3].apply(lambda x:np.sum(x == 0))
print(count_zero)

var3                     75
var15                     0
imp_ent_var16_ult1    72301
dtype: int64


### 3-2 描述统计-均值和中位数

#### 4 均值 `Mean Values`

In [58]:
np.mean(df.iloc[:, 0]) # 没有去除缺失值之前的均值很低

-1523.1992765061825

去除缺失值, 原数据 `76020` 行, 去除无效的数据后为 `75904`.<br>
`np.isin(a, b, invert=True)`<br>
` a `中的元素是否在` b `中, 如果存在返回` True `, 不存在返回` False `.<br>
如果设置了`invert=True`,则情况恰恰相反，即` a `中元素在` b `中则返回` False`

In [56]:
df.iloc[:, 0][np.isin(df.iloc[:, 0], missSet, invert=True)].shape

(75904,)

对过滤后的数据进行均值运算

In [57]:
np.mean(df.iloc[:, 0][np.isin(df.iloc[:, 0], missSet, invert=True)])

2.717577466273187

In [63]:
df_mean = df.iloc[:, 0:3].apply(lambda x: np.mean(x[np.isin(x, missSet, invert=True)]))
print(df_mean)

var3                   2.717577
var15                 33.212865
imp_ent_var16_ult1    86.208265
dtype: float64


#### 5 中位数 `Median Values`

In [65]:
np.median(df.iloc[:, 0])  # 没有去除缺失值之前的均值很低

2.0

In [68]:
np.median(df.iloc[:, 0][np.isin(df.iloc[:, 0], missSet, invert=True)]) # 去除之后的第一列中指

2.0

In [71]:
df_median = df.iloc[:, 0:3].apply(lambda x: np.median(x[np.isin(x, missSet, invert=True)]))
print(df_median)

var3                   2.0
var15                 28.0
imp_ent_var16_ult1     0.0
dtype: float64


### 3-3 统计描述-众数

#### 6 众数 `Mode Values`

<font face="Courier New">

`numpy` 中没有求众数的这一方法, 因此使用 `scipy` 中的 `state` 方法.<br>
该方法返回两个值, 第一个为众数本身(list), 第二个数为众数出现的次数(list).

In [101]:
stats.mode(df.iloc[:, 0][np.isin(df.iloc[:, 0], missSet, invert=True)])

ModeResult(mode=array([2]), count=array([74165]))

In [89]:
df_mode = df.iloc[:, 0:3].apply(lambda x: stats.mode(x[np.isin(x, missSet, invert=True)])[0][0])
print(df_mode)

var3                   2.0
var15                 23.0
imp_ent_var16_ult1     0.0
dtype: float64


In [103]:
df_mode_count = df.iloc[:, 0:3].apply(lambda x: stats.mode(x[np.isin(x, missSet, invert=True)])[1][0])
print(df_mode_count)

var3                  74165
var15                 20170
imp_ent_var16_ult1    72301
dtype: int64


In [106]:
df_mode_perct = df_mode_count / df.shape[0] # 每一列的众数所占的比例
print(df_mode_perct)

var3                  0.975599
var15                 0.265325
imp_ent_var16_ult1    0.951079
dtype: float64


### 3-4 描述统计-最大最小值

#### 7 最大值 `Max Values`

In [108]:
np.max(df.iloc[:, 0]) # 没有去除缺失值之前

238

In [109]:
np.max(df.iloc[:, 0:3].apply(lambda x: x[np.isin(x, missSet, invert=True)]))

var3                     238.0
var15                    105.0
imp_ent_var16_ult1    210000.0
dtype: float64

#### 8 最小值 `Min Values`

In [107]:
np.min(df.iloc[:, 0])

-999999

In [30]:
np.min(df.iloc[:, 0:3].apply(lambda x: x[np.isin(x, missSet, invert=True)]))

var3                  0.0
var15                 5.0
imp_ent_var16_ult1    0.0
dtype: float64

### 3-5 统计描述-分位点

#### 9 分位点 `quantile Values`

<font face="Courier New">

`np.percentile(a, q, axis, keepdims)`
- a : np数组
- q : `float in range of [0,100] (or sequence of floats)` Percentile to compute, 要计算的q分位数。
- axis : 那个轴上运算。
- keepdims : bool是否保持维度不变。

In [28]:
ddff = pd.Series(np.arange(10, 21), index=["a","b","c","d","e","f","g","h","i","j","k"])
print(ddff)
np.percentile(ddff, (0, 20, 50, 70, 100)) # 后面的元组是百分比

a    10
b    11
c    12
d    13
e    14
f    15
g    16
h    17
i    18
j    19
k    20
dtype: int64


array([10., 12., 15., 17., 20.])

In [36]:
np.percentile(df.iloc[:, 0], (1, 5, 25, 50, 75, 95, 99)) # 去除缺失值之前

array([ 2.,  2.,  2.,  2.,  2.,  2., 11.])

In [37]:
np.percentile(df.iloc[:, 0][np.isin(df.iloc[:, 0], missSet, invert=True)], (1, 5, 25, 50, 75, 95, 99))

array([ 2.,  2.,  2.,  2.,  2.,  2., 11.])

<font face="Courier New">

`enumerate()` 函数用于将一个可遍历的数据对象(如列表、元组或字符串)组合为一个索引序列，同时列出数据和数据下标，一般用在 for 循环当中。<br>

语法: `enumerate(sequence, [start=0])`<br>
参数: `sequence` -- 一个序列、迭代器或其他支持迭代对象。  `start` -- 下标起始位置。<br>
返回值: 返回 `enumerate(枚举)` 对象。

求出前三列的分位数

In [38]:
json_percentile = {}
for i, name in enumerate(df.iloc[:, 0:3].columns):
    json_percentile[name] = np.percentile(df.iloc[:, i][np.isin(df.iloc[:, i], missSet, invert=True)], (1, 5, 25, 50, 75, 95, 99))

print(json_percentile)

{'var3': array([ 2.,  2.,  2.,  2.,  2.,  2., 11.]), 'var15': array([22., 23., 23., 28., 40., 60., 77.]), 'imp_ent_var16_ult1': array([   0.,    0.,    0.,    0.,    0.,    0., 1500.])}


将生成的 `json_percentile` 转为`Series`

In [51]:
pd.DataFrame(json_percentile)[['imp_ent_var16_ult1', 'var3', 'var15']].T  # 根据列名索引的顺序不同, 结果顺序也不同

0     1     2     3     4     5       6
imp_ent_var16_ult1   0.0   0.0   0.0   0.0   0.0   0.0  1500.0
var3                 2.0   2.0   2.0   2.0   2.0   2.0    11.0
var15               22.0  23.0  23.0  28.0  40.0  60.0    77.0

In [52]:
pd.DataFrame(json_percentile)[df.iloc[:, 0:3].columns].T

0     1     2     3     4     5       6
var3                 2.0   2.0   2.0   2.0   2.0   2.0    11.0
var15               22.0  23.0  23.0  28.0  40.0  60.0    77.0
imp_ent_var16_ult1   0.0   0.0   0.0   0.0   0.0   0.0  1500.0

### 3-6 描述统计-频数

#### 10 频数

取出第一列的前五行的频数, `value_counts` 会将各个频数按大到小的顺序排列.

In [59]:
df.iloc[:, 0].value_counts().iloc[0:5, ]

 2         74165
 8           138
-999999      116
 9           110
 3           108
Name: var3, dtype: int64

去除缺失值之后的第一列前五行频数

In [75]:
df.iloc[:, 0][np.isin(df.iloc[:, 0], missSet, invert=True)].value_counts().iloc[0:5, ]

2    74165
8      138
9      110
3      108
1      105
Name: var3, dtype: int64

In [76]:
print(df.iloc[:, 3][np.isin(df.iloc[:, 3], missSet, invert=True)].value_counts().iloc[0:5, ])
print(df.iloc[:, 2][np.isin(df.iloc[:, 2], missSet, invert=True)].value_counts().iloc[0:5, ])

0.0     66075
30.0      120
60.0      100
15.0       56
90.0       48
Name: imp_op_var39_comer_ult1, dtype: int64
0.0      72301
300.0      221
150.0      217
600.0      172
900.0      164
Name: imp_ent_var16_ult1, dtype: int64


将 `imp_op_var39_comer_ult1` 与 `imp_ent_var16_ult1` 两列合并

In [104]:
json_fre_name = {}
json_fre_count = {}

def fill_fre_top_5(x):
    if len(x) < 7:
        new_array = np.full(7, np.nan)
        new_array[0: len(x)] = x
        return new_array

for i, name in enumerate(df[["imp_op_var39_comer_ult1", "imp_ent_var16_ult1"]]):
    index_name = df[name][np.isin(df[name], missSet, invert=True)].value_counts().iloc[0:5,].index.values
    index_name = fill_fre_top_5(index_name)
    json_fre_name[name] = index_name
    
    value_count = df[name][np.isin(df[name], missSet, invert=True)].value_counts().iloc[0:5,].values
    value_count = fill_fre_top_5(value_count)
    json_fre_count[name] = value_count

df_pre_name = pd.DataFrame(json_fre_name)[["imp_op_var39_comer_ult1", "imp_ent_var16_ult1"]].T
df_pre_count = pd.DataFrame(json_fre_count)[["imp_op_var39_comer_ult1", "imp_ent_var16_ult1"]].T
df_pre = pd.concat([df_pre_name, df_pre_count], axis=1)
print(df_pre)

                           0      1      2      3      4   5   6        0  \
imp_op_var39_comer_ult1  0.0   30.0   60.0   15.0   90.0 NaN NaN  66075.0   
imp_ent_var16_ult1       0.0  300.0  150.0  600.0  900.0 NaN NaN  72301.0   

                             1      2      3      4   5   6  
imp_op_var39_comer_ult1  120.0  100.0   56.0   48.0 NaN NaN  
imp_ent_var16_ult1       221.0  217.0  172.0  164.0 NaN NaN  


### 3-7 统计描述-缺失值

#### 11 缺失值 `Miss Values`

In [107]:
np.sum(np.isin(df.iloc[:, 0], missSet)) # 存在116个缺失值

116

In [115]:
sd_miss = df.iloc[:, 0:3].apply(lambda x: np.sum(np.isin(x, missSet)))
print(sd_miss)

var3                  116
var15                   0
imp_ent_var16_ult1      0
dtype: int64


## 四 数据诊断工具验证

### 4-1 工具整合

In [123]:
def eda_analysis(missSet=[np.nan, 9999999999, -999999], dataframe=None):
    # 1Count
    count_un = df.apply(lambda x: len(x.unique()))
    count_un = count_un.to_frame('count')

    # 2Zero Values #
    count_zero = df.apply(lambda x: np.sum(x == 0))
    count_zero = count_zero.to_frame('count_zero')

    # 3 Mean #
    df_mean = df.apply(lambda x: np.mean(x[np.isin(x, missSet, invert=True)]))
    df_mean = df_mean.to_frame('mean')

    # 4 Median #
    df_median = df.apply(lambda x: np.median(x[np.isin(x, missSet, invert=True)]))
    df_median = df_median.to_frame('median')

    # 5 Mode #
    df_mode = df.apply(lambda x: stats.mode(x[np.isin(x, missSet, invert=True)])[0][0])
    df_mode = df_mode.to_frame('mode')

    # 6 Mode Percentage#
    df_mode_count = df.apply(lambda x: stats.mode(x[np.isin(x, missSet, invert=True)])[1][0])
    df_mode_count = df_mode_count.to_frame('mode_count')

    df_mode_perct = df_mode_count / df.shape[0]
    df_mode_perct.columns = ['mode_perct']

    # 7 Min Value#
    df_min = df.apply(lambda x: np.min(x[np.isin(x, missSet, invert=True)]))
    df_min = df_min.to_frame('min')

    # 8 Max Value#
    df_max = df.apply(lambda x: np.max(x[np.isin(x, missSet, invert=True)]))
    df_max = df_max.to_frame('max')

    # 9 Quantile Values#
    json_quantile = {}
    for i, name in enumerate(df.columns):
        json_quantile[name] = np.percentile(df[name][np.isin(df[name], missSet, invert=True)],
                                            (1, 5, 25, 50, 75, 95, 99))
    df_quantile = pd.DataFrame(json_quantile)[df.columns].T
    df_quantile.columns = ['quan01', 'quan05', 'quan25', 'quan50', 'quan75', 'quan95', 'quan99']

    # 10 Frequence
    json_fre_name = {}
    json_fre_count = {}

    for i, name in enumerate(df.columns):
        index_name = df[name][np.isin(df[name], missSet, invert=True)].value_counts().iloc[0:5, ].index.values
        index_name = fill_fre_top_5(index_name)
        json_fre_name[name] = index_name

        value_count = df[name][np.isin(df[name], missSet, invert=True)].value_counts().iloc[0:5, ].values
        value_count = fill_fre_top_5(value_count)
        json_fre_count[name] = value_count

    df_fre_name = pd.DataFrame(json_fre_name)[df.columns].T
    df_fre_count = pd.DataFrame(json_fre_count)[df.columns].T
    df_fre = pd.concat([df_fre_name, df_fre_count], axis=1)
    df_fre.colums = ["value1", "value2", "value3", "value4", "value5", "freq1", "freq2", "freq3", "freq4", "freq5"]

    # 11 Miss Value
    df_miss = df.apply(lambda x: np.sum(np.isin(x, missSet)))
    df_miss = df_miss.to_frame("freq_miss")

    # 12 Combine all informations
    df_eda_summary = pd.concat([
        count_un, count_zero, df_mean, df_median, df_mode, df_mode_count, df_mode_perct,
        df_min, df_max, df_fre, df_miss
    ], axis=1)

    return df_eda_summary

In [124]:
eda_analysis(missSet = [np.nan, 9999999999, -999999], dataframe=df.iloc[:, 0:3])

/home/ubuntu/Env/py3tf/lib/python3.6/site-packages/ipykernel_launcher.py:61: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access


count  count_zero           mean     median  \
var3                             208          75       2.717577       2.00   
var15                            100           0      33.212865      28.00   
imp_ent_var16_ult1               596       72301      86.208265       0.00   
imp_op_var39_comer_ult1         7551       66075      72.363067       0.00   
imp_op_var39_comer_ult3         9099       64330     119.529632       0.00   
imp_op_var40_comer_ult1          293       75725       3.559130       0.00   
imp_op_var40_comer_ult3          346       75672       6.472698       0.00   
imp_op_var40_efect_ult1           23       75982       0.412946       0.00   
imp_op_var40_efect_ult3           29       75975       0.567352       0.00   
imp_op_var40_ult1                224       75791       3.160715       0.00   
imp_op_var41_comer_ult1         7421       66226      68.803937       0.00   
imp_op_var41_comer_ult3         8961       64480     113.056934       0.00   
imp_op_var41_efect_ult1          331       69336      68.205140       0.00   
imp_op_var41_efect_ult3          454       67372     113.225058       0.00   
imp_op_var41_ult1               8032       64388     137.242763       0.00   
imp_op_var39_efect_ult1          336       69325      68.618087       0.00   
imp_op_var39_efect_ult3          462       67360     113.792410       0.00   
imp_op_var39_ult1               8149       64271     140.403479       0.00   
imp_sal_var16_ult1                66       75924       5.477676       0.00   
ind_var1_0                         2       75149       0.011458       0.00   
ind_var1                           2       75734       0.003762       0.00   
ind_var2_0                         1       76020       0.000000       0.00   
ind_var2                           1       76020       0.000000       0.00   
ind_var5_0                         2        3191       0.958024       1.00   
ind_var5                           2       25561       0.663760       1.00   
ind_var6_0                         2       76012       0.000105       0.00   
ind_var6                           2       76018       0.000026       0.00   
ind_var8_0                         2       73524       0.032833       0.00   
ind_var8                           2       73846       0.028598       0.00   
ind_var12_0                        2       70887       0.067522       0.00   
...                              ...         ...            ...        ...   
saldo_medio_var12_ult3          3447       72544    4401.002439       0.00   
saldo_medio_var13_corto_hace2   1628       73047    3639.419939       0.00   
saldo_medio_var13_corto_hace3    968       74657     556.184178       0.00   
saldo_medio_var13_corto_ult1     943       72836    4852.261814       0.00   
saldo_medio_var13_corto_ult3    2576       72836    3857.848542       0.00   
saldo_medio_var13_largo_hace2    295       75532     771.227449       0.00   
saldo_medio_var13_largo_hace3    264       75722     162.170439       0.00   
saldo_medio_var13_largo_ult1     233       75483     956.950207       0.00   
saldo_medio_var13_largo_ult3     511       75483     750.956273       0.00   
saldo_medio_var13_medio_hace2      3       76018       0.175324       0.00   
saldo_medio_var13_medio_hace3      1       76020       0.000000       0.00   
saldo_medio_var13_medio_ult1       3       76018       0.513023       0.00   
saldo_medio_var13_medio_ult3       3       76018       0.344174       0.00   
saldo_medio_var17_hace2           88       75930      91.171811       0.00   
saldo_medio_var17_hace3           18       76003      36.463184       0.00   
saldo_medio_var17_ult1           119       75902     131.031566       0.00   
saldo_medio_var17_ult3           119       75902     109.216944       0.00   
saldo_medio_var29_hace2            5       76016       0.213071       0.00   
saldo_medio_var29_hace3            2       76019       0.001910       0.00   
saldo_medio_var29_ult1             4       76017     

### 4-2 工具性能和效率测试